[SaferCPP] Wrap FrameSelection in CheckedRef

atar13 opened the pull request:
#### 9a029b1e8b76e34c2103c203e635c17e20aaa800
<pre>
[SaferCPP] Wrap FrameSelection in CheckedRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=300534">https://bugs.webkit.org/show_bug.cgi?id=300534</a>
<a href="https://rdar.apple.com/162403789">rdar://162403789</a>

Reviewed by Ryosuke Niwa.

Wrap usage of FrameSelection in CheckedRef to check the lifetime
before it gets used as the `this` parameter of `textWasReplaced()`.
Fixes static analyzer warning:
 &quot;Call argument for &apos;this&apos; parameter is unchecked and unsafe&quot;.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::CharacterData::setData):
(WebCore::CharacterData::setDataAndUpdate):

Canonical link: <a href="https://commits.webkit.org/301499@main">https://commits.webkit.org/301499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/781310667410f216bcef0c22de66c9dd61296626

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125946 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45607 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132813 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77803 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ab739fa2-fe02-421f-9915-5891e5acb352) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46290 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95953 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64049 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9fec63c1-fb5c-4da3-b04d-f04bc38796c3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112630 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76442 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/96454ef8-9934-4cb2-a133-e88972d8e8fd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35929 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30816 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76283 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106803 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31028 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135499 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52729 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40459 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104425 "") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53178 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104145 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26578 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27845 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50107 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52623 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58437 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51967 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55315 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53664 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/CharacterData.cpp:
```diff
@@ -65,8 +65,10 @@ void CharacterData::setData(const String& data)
-        if (RefPtr frame = document->frame())
-            frame->selection().textWasReplaced(*this, 0, oldLength, oldLength);
+        if (RefPtr frame = document->frame()) {
+            CheckedRef protectedSelection = frame->selection();
```
We should add `Frame::checkedSelection()` which returns `CheckedRef<FrameSelection>` and call that here instead of allocating a local variable.

atar13 replied:
`LocalFrame` actually already had a `checkedSelection()`, so switching to that.

rniwa commented on Source/WebCore/dom/CharacterData.cpp:
```diff
@@ -205,8 +207,10 @@ void CharacterData::setDataAndUpdate(const String& newData, unsigned offsetOfRep
-    if (RefPtr frame = document->frame())
-        frame->selection().textWasReplaced(*this, offsetOfReplacedData, oldLength, newLength);
+    if (RefPtr frame = document->frame()) {
+        CheckedRef protectedSelection = frame->selection();
```
Ditto.

atar13 committed:
[SaferCPP] Wrap FrameSelection in CheckedRef
https://bugs.webkit.org/show_bug.cgi?id=300534
rdar://162403789

Reviewed by Ryosuke Niwa.

Wrap usage of FrameSelection in CheckedRef to check the lifetime
before it gets used as the `this` parameter of `textWasReplaced()`.
Fixes static analyzer warning:
 "Call argument for 'this' parameter is unchecked and unsafe".

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::CharacterData::setData):
(WebCore::CharacterData::setDataAndUpdate):

Canonical link: https://commits.webkit.org/301499@main
